tracing/trace_state.cc

```diff
@@ -144,20 +148,18 @@ void trace_state::set_user_timestamp(api::timestamp_type val) {
 }
 
 void trace_state::add_prepared_statement(prepared_checked_weak_ptr& prepared) {
-    _params_ptr->prepared_statements.emplace_back(prepared->checked_weak_from_this());
+    _params_ptr->prepared_statements.push_back(params_values::prepared_statement_info{prepared->checked_weak_from_this(), {}, {}});
```

Comment:
I started with `emplace_back`, but it failed to deduce the types from `{}` passed as second and third parameters, and the definitions of these types are really long and I'd rather not explicitly copy them. So, I decided to push_back, since it's likely going to be compiled to identical code anyway.

Reply:
I see. Could you, please, add an inline comment above this line then so that we would understand why it's this way in the future. ;)

Reply:
I don't follow what would be the contents of such a comment, or its purpose. In my opinion, comments should describe non-obvious semantics, which is not the case here - it's clear that we're just putting stuff into a vector. Both emplace_back and push_back makes perfect sense here, it's just that the code will get longer in case of emplace_back due to having to explicitly state all the types. It's a kind of a question that can be answered via review, and doesn't warrant a comment in the code. We already have multiple precedents in the code similar to this one, none of them are commented:
```
[sarna@localhost scylla]$ git grep push_back.*\,\ \{\}
clustering_ranges_walker.hh:                    rts.push_back(range_tombstone_change(_current_end, {}));
raft/server.cc:    _reads.push_back({rid->first, rid->second, {}});
repair/row_level.cc:                            row_list.push_back(repair_row({}, {}, dk_ptr, {}, is_dirty_on_master::no, std::move(mf)));
sstables/mutation_fragment_filter.hh:                pr.rts.push_back(range_tombstone_change(_fwd_end, {}));
utils/logalloc.cc:        _backtraces.push_back(std::make_unique<backtrace_entry>(backtrace_entry{current_backtrace(), {}}));
```

Reply:
Well, using `emplace` makes more sense here since you are constructing the item in-place.
However you DO have a good-enough reason to use push_back which is not obvious unless you explain it.

So, the comment could be something like that: 
```
// Use push_back() instead of emplace_back() to avoid long explicit casts
```

Reply:
I think the initialization is somewhat bogus: you initialize an `optional<vector>` with an empty vector which means that until it's reassigned with an actual value it would be holding an actual value (empty vector) which is both logically confusing (it's supposed to be uninitialized) but also inefficient because that empty vector value is not needed when unnamed prepared parameters semantics is used (most common one) and then this vector value is going to be discarded the moment this field is going to be assigned `nullopt`.

Long story short, it would be better if you would implement that constructor I mentioned above. This would also allow you to use `emplace_back()` here and keep the code line both short and self-explanatory.

Reply:
It's not bogus - intializing an `optional<T>` with `{}` is equivalent to initializing it with a disengaged optional, just like passing `std::nullopt`. No empty vector would be constructed and there's no inefficiency. Regardless, I don't mind creating a custom constructor if it makes the review process go forward, I'll push an update next week.